src/main/java/net/glowstone/util/pattern/PortalShape.java

```diff
@@ -0,0 +1,248 @@
+            // move downwards
+            location.subtract(0, 1, 0);
+        }
+
+        // No obsidian in 23 block range
+        if (downImmutable(location).getBlock().getType() != Material.OBSIDIAN) {
+            return;
+        }
+
+        // get distance to leftmost edge
+        int i = getDistanceUntilEdge(location, left);
+
+        // No edge -> no portal
+        if (i < 0) {
+            return;
+        }
+
+        // Set the bottom leftmost portal bloc
+        bottomLeft = offsetImmutable(location, left, i);
+
+        // Calculate the portal width
+        width = getDistanceUntilEdge(bottomLeft, left.getOppositeFace()) + 1;
+
+        // Portal too big / too small
+        if (width < 2 || width > 21) {
+            return;
+        }
+
+        // calculate portal height
+        height = calculatePortalHeight();
+    }
+
+    /**
+     * Check whether a given Material counts as empty in a portal.
+     *
+     * @param material the material to check for
+     * @return whether the material counts as empty
+     */
+    private static boolean canBuildThrough(Material material) {
+        return material == Material.AIR || material == Material.FIRE || material == Material.PORTAL;
+    }
+
+    /**
+     * Go down 1 block from a Location without changing the Location itself.
+     * Needed to perform a "look-ahead" while preserving the original location.
+     *
+     * @param in the location to go down from
+     * @return a new Location with the y-value decreased by one
+     */
+    private static Location downImmutable(Location in) {
+        return new Location(in.getWorld(), in.getX(), in.getY() - 1, in.getZ());
+    }
+
+    /**
+     * Offset from a given location in a certain direction by a specific amount of times.
+     * Needed to preserve the original location while offsetting.
+     *
+     * @param in     the original location to offset from
+     * @param face   the direction in which to offset
+     * @param offset how many times to offset
+     * @return a new Location with the specified offset
+     */
+    private static Location offsetImmutable(Location in, BlockFace face, int offset) {
+        return new Location(
+            in.getWorld(),
+            in.getX() + (face.getModX() * offset),
+            in.getY() + (face.getModY() * offset),
+            in.getZ() + (face.getModZ() * offset)
+        );
+    }
+
+    /**
+     * Get the distance to one edge of the possible portal. Returns -1 if no end is reached.
+     *
+     * @param location the location from where to check. Not modified in process.
+     * @param face     the side of the edge to check for
+     * @return the distance or -1 in case of an invalid shape
+     */
+    private int getDistanceUntilEdge(Location location, BlockFace face) {
+        // Clone the location as Location objects are mutable
+        Location destLoc = location.clone();
+
+        int i;
```

Comment:
Rename `i` to `distance` in this method.

Reply:
Applied